File: src/hello_world_service_py/hello_world_service_py/async_client.py
#!/usr/bin/env python3
import sys
import asyncio
import threading
import rclpy
from rclpy.node import Node
from example_interfaces.srv import AddTwoInts

class AddTwoIntsClientAsync(Node):
    def __init__(self):
        super().__init__("add_two_ints_client_async")
        self.client = self.create_client(AddTwoInts, "add_two_ints")
        
        while not self.client.wait_for_service(timeout_sec=1.0):
            self.get_logger().info("服务未启动，等待中...")
            if not rclpy.ok():
                self.get_logger().error("节点被中断，退出")
                return
        
        self.get_logger().info("服务已连接！")
    
    async def send_request(self, a: int, b: int):
        """异步发送请求并等待响应"""
        request = AddTwoInts.Request()
        request.a = a
        request.b = b
        
        # 调用服务（返回ROS2 Future）
        ros_future = self.client.call_async(request)
        self.get_logger().info(f"发送请求: {a} + {b}")
        
        # 将ROS2 Future转换为asyncio Future
        loop = asyncio.get_running_loop()
        asyncio_future = loop.create_future()
        
        def done_callback(rf):
            """ROS2服务完成时的回调"""
            try:
                result = rf.result()
                loop.call_soon_threadsafe(asyncio_future.set_result, result)
            except Exception as e:
                loop.call_soon_threadsafe(asyncio_future.set_exception, e)
        
        # 添加回调
        ros_future.add_done_callback(done_callback)
        
        # 异步等待结果
        response = await asyncio_future
        self.get_logger().info(f"收到响应: {response.sum}")
        return response.sum


async def async_main(args=None):
    """异步主函数"""
    rclpy.init(args=args)
    
    # 检查参数
    if len(sys.argv) != 3:
        print("用法: ros2 run 包名 add_two_ints_client_async <a> <b>")
        rclpy.shutdown()
        return
    
    try:
        a = int(sys.argv[1])
        b = int(sys.argv[2])
    except ValueError:
        print("参数必须是整数！")
        rclpy.shutdown()
        return
    
    # 创建客户端
    client = AddTwoIntsClientAsync()
    
    # 在后台线程中运行ROS2 spin
    def spin():
        rclpy.spin(client)
    
    spin_thread = threading.Thread(target=spin, daemon=True)
    spin_thread.start()
    
    try:
        # 异步调用服务
        result = await client.send_request(a, b)
        print(f"\n最终结果: {a} + {b} = {result}")
    finally:
        rclpy.shutdown()


def main():
    asyncio.run(async_main())


if __name__ == "__main__":
    main()

File: src/hello_world_service_cpp/src/async_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/srv/add_two_ints.hpp"
#include "hello_world_service_cpp/async_client.h"

using namespace std::chrono_literals;

AddTwoIntsAsyncClient::AddTwoIntsAsyncClient() : Node("async_add_two_ints_client")
{
    client_ = this->create_client<example_interfaces::srv::AddTwoInts>("/add_two_ints");
}

AddTwoIntsAsyncClient::~AddTwoIntsAsyncClient()
{
}

bool AddTwoIntsAsyncClient::add_two_ints(int a, int b)
{
    // 等待服务端
    if (!client_->wait_for_service(5s))
    {
        RCLCPP_ERROR(this->get_logger(), "等待服务端超时");

        return false;
    }

    auto request = std::make_shared<AddTwoInts::Request>();
    request->a = a;
    request->b = b;

    RCLCPP_INFO(this->get_logger(), "发送请求: %ld + %ld", request->a, request->b);

    auto result = client_->async_send_request(request,
            std::bind(&AddTwoIntsAsyncClient::response_callback, this, std::placeholders::_1));

    return true;
}

void AddTwoIntsAsyncClient::response_callback(rclcpp::Client<example_interfaces::srv::AddTwoInts>::SharedFuture future)
{
    try
    {
        auto response = future.get();
        RCLCPP_INFO(this->get_logger(), "异步回调收到响应: %ld", response->sum);
    }
    catch (const std::exception &e)
    {
        RCLCPP_ERROR(this->get_logger(), "服务调用失败: %s", e.what());
    }
}

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<AddTwoIntsAsyncClient>();

    if (argc >= 3)
    {
        int a = std::stoi(argv[1]);
        int b = std::stoi(argv[2]);
        node->add_two_ints(a, b);
    }

    rclcpp::spin(node);
    rclcpp::shutdown();

    return 0;
}